src/crystal/static_list.cr

```diff
@@ -0,0 +1,97 @@
+# :nodoc:
+struct Crystal::StaticList
```

Comment:
I would like to find a more crystal way to build this static list.
I think it should be possible to use a generic struct that will enforce typing and maybe allow some macros to hide the boilerplate as done here. But all the pointer, casting and final interface make me doubt.

Also, although it `:nodoc:` I think there could be some specs on the implementation.

The main problem is to build a linked list with data that should act as nodes and let them be stored in the stack, right?

Reply:
I think this is a difference between [internal storage and external storage](https://en.wikipedia.org/wiki/Linked_list#Internal_and_external_storage). The external storage will enforce type explicit, and the internal storage requires some way to know which type stores the list node. That's why need the `Crystal::StaticList.container_of`. Because we store data in stack, it cannot prevent to handle without pointer.

And I will add some spec about the `StaticList`.

Reply:
What do you think about something like this https://gist.github.com/bcardiff/4327823bb43cbb93734fbb6240315dac ?

* there are less casts
* the pointer tricker is hidden in the module
* the API is a bit more clean: sometimes as methods, sometimes as macro calls.

there might be some subtle differences regarding the representation of the list. and this solution can only be used in `structs`. I didn't care to apply them to classes since I think it is only used in structs.

In case we need to extend it, in the worst case macros can be injected in the type that is including the module to specialize if the next/prev need to be pointer(T) or T.

Reply:
Sorry for late on reply because I'm busy today. In my opinion, this change is a bit worse than mine. `Listify.prepend(list, node)` will make code like c, not crystal. Although I currently use `node.append_to pointerof(list)` in code, the implementation can be easily replaced by `list.append node.self_pointer`. `list.append` and `list.shift` will be better than `Listify.append` and `Listify.shift`.

Besides, the `each` operation should not hide the pointer. If you hide the pointer, the object you get is the copy object of the pointer point to. Any changes will not apply on the original object. This will be a problem.

Including a module is good, but I don't want to limit the name of member variable in  struct/class which use the list. (If possible, they may need the list with different name for future refactor/optimize)

Reply:
Look at the way that `Thread::LinkedList` does it, and i'm pretty sure you can generalise that to work on the raw pointers. Then you don't need the `init_empty_list` either, just store the head and tail and use null pointers to signify the start/end of the list.